File: examples/python/megacli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Simple command line interface using the Python bindings."""

## Created: 15 Jan 2015 Javier Serrano <js@mega.co.nz>
##
## (c) 2015 by Mega Limited, Auckland, New Zealand
##     http://mega.co.nz/
##     Simplified (2-clause) BSD License.
##
## You should have received a copy of the license along with this
## program.
##
## This file is part of the multi-party chat encryption suite.
##
## This code is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

__author__ = 'Javier Serrano <js@mega.co.nz>'

import sys
import os
import cmd
import logging
import time

from mega import (MegaApi, MegaListener, MegaError, MegaRequest,
                  MegaUser, MegaNode)


class AppListener(MegaListener):
    def __init__(self, shell):
        self._shell = shell
        super(AppListener, self).__init__()


    def onRequestStart(self, api, request):
        logging.info('Request start ({})'.format(request))


    def onRequestFinish(self, api, request, error):
        logging.info('Request finished ({}); Result: {}'
                     .format(request, error))
        if error.getErrorCode() != MegaError.API_OK:
            return

        request_type = request.getType()
        if request_type == MegaRequest.TYPE_LOGIN:
            api.fetchNodes()
        elif request_type == MegaRequest.TYPE_EXPORT:
            logging.info('Exported link: {}'.format(request.getLink()))
        elif request_type == MegaRequest.TYPE_ACCOUNT_DETAILS:
            account_details = request.getMegaAccountDetails()
            logging.info('Account details received')
            logging.info('Account e-mail: {}'.format(api.getMyEmail()))
            logging.info('Storage: {} of {} ({} %)'
                         .format(account_details.getStorageUsed(),
                                 account_details.getStorageMax(),
                                 100 * account_details.getStorageUsed()
                                 / account_details.getStorageMax()))
            logging.info('Pro level: {}'.format(account_details.getProLevel()))


    def onRequestTemporaryError(self, api, request, error):
        logging.info('Request temporary error ({}); Error: {}'
                .format(request, error))


    def onTransferFinish(self, api, transfer, error):
        logging.info('Transfer finished ({}); Result: {}'
                .format(transfer, transfer.getFileName(), error))


    def onTransferUpdate(self, api, transfer):
        logging.info('Transfer update ({} {});'
                     ' Progress: {} KB of {} KB, {} KB/s'
                     .format(transfer,
                             transfer.getFileName(),
                             transfer.getTransferredBytes() / 1024,
                             transfer.getTotalBytes() / 1024,
                             transfer.getSpeed() / 1024))


    def onTransferTemporaryError(self, api, transfer, error):
        logging.info('Transfer temporary error ({} {}); Error: {}'
                     .format(transfer, transfer.getFileName(), error))


    def onUsersUpdate(self, api, users):
        if users != None:
            logging.info('Users updated ({})'.format(users.size()))

    def onNodesUpdate(self, api, nodes):
        if nodes != None:
            logging.info('Nodes updated ({})'.format(nodes.size()))
        else:
            self._shell.cwd = api.getRootNode()


class MegaShell(cmd.Cmd, MegaListener):
    intro = 'Mega sample app. Type help or ? to list commands.\n'
    PROMPT = '(MEGA)'

    def __init__(self, api):
        self._api = api
        self.cwd = None
        super(MegaShell, self).__init__()


    def emptyline(self):
        return


    def do_login(self, arg):
        """Usage: login email password"""
        args = arg.split()
        if len(args) != 2 or '@' not in args[0]:
            print(self.do_login.__doc__)
            return

        self._api.login(args[0], args[1])


    def do_logout(self, arg):
        """Usage: logout"""

        args = arg.split()
        if len(args) != 0:
            print(self.do_logout.__doc__)
            return

        if self.cwd == None:
            print('INFO: Not logged in')
            return

        self._api.logout()
        self.cwd = None


    def do_mount(self, arg):
        """Usage: mount"""
        args = arg.split()
        if len(args) != 0:
            print(self.do_mount.__doc__)
            return
        if not self._api.isLoggedIn():
            print('INFO: Not logged in')
            return

        print('INFO: INSHARES:')
        users = self._api.getContacts()
        for i in range(users.size()):
            user = users.get(i)
            if user.getVisibility() == MegaUser.VISIBILITY_VISIBLE:
                shares = self._api.getInShares(user)
                for j in range(shares.size()):
                    share = shares.get(j)
                    print('INFO: INSHARE on {} {} Access level: {}'
                          .format(users.get(i).getEmail(),
                                  are.getName(),
                                  self._api.getAccess(share)))


    def do_ls(self, arg):
        """Usage: ls [path]"""
        args = arg.split()
        if len(args) > 1:
            print(self.do_ls.__doc__)
            return

        if self.cwd == None:
            print('INFO: Not logged in')
            return

        path = None
        if len(args) == 0:
            path = self.cwd
        else:
            path = self._api.getNodeByPath(args[0], self.cwd)

        print('    .')
        if self._api.getParentNode(path) != None:
            print('    ..')

        nodes = self._api.getChildren(path)
        for i in range(nodes.size()):
            node = nodes.get(i)
            output = '    {}'.format(node.getName())
            if node.getType() == MegaNode.TYPE_FILE:
                output += '   ({} bytes)'.format(node.getSize())
            else:
                output += '   (folder)'
            print(output)


    def do_cd(self, arg):
        """Usage: cd [path]"""
        args = arg.split()
        if len(args) > 1:
            print(self.do_cd.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        if len(args) == 0:
            self.cwd = self._api.getRootNode()
            return

        node = self._api.getNodeByPath(args[0], self.cwd)
        if node == None:
            print('{}: No such file or directory'.format(args[0]))
            return
        if node.getType() == MegaNode.TYPE_FILE:
            print('{}: Not a directory'.format(args[0]))
            return
        self.cwd = node


    def do_get(self, arg):
        """Usage: get remotefile"""
        args = arg.split()
        if len(args) != 1:
            print(self.do_get.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        node = self._api.getNodeByPath(args[0], self.cwd)
        if node == None:
            print('Node not found: {}'.format(args[0]))
            return

        self._api.startDownload(node, './', None, None, False, None, MegaTransfer.COLLISION_CHECK_FINGERPRINT, MegaTransfer.COLLISION_RESOLUTION_NEW_WITH_N)


    def do_put(self, arg):
        """Usage: put localfile"""
        args = arg.split()
        if len(args) != 1:
            print(self.do_put.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        self._api.startUpload(args[0], self.cwd, None, 0, None, False, False, None)


    def do_mkdir(self, arg):
        """Usage: mkdir path"""
        args = arg.split()
        if len(args) != 1:
            print(self.do_mkdir.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        base = self.cwd
        name = args[0]
        if '/' in name or '\\' in name:
            index1 = name.rfind('/')
            index2 = name.rfind('\\')
            index = None
            if index1 > index2:
                index = index1
            else:
                index = index2
            path = name[:index + 1]
            base = self._api.getNodeByPath(path, self.cwd)
            name = name[index + 1:]

            if not name:
                print('{}: Path already exists'.format(path))
                return
            if base == None:
                print('{}: Target path not found'.format(path))
                return

        check = self._api.getNodeByPath(name, base)
        if check != None:
            print('{}: Path already exists'
                  .format(self._api.getNodePath(check)))
            return

        self._api.createFolder(name, base)


    def do_rm(self, arg):
        """Usage: rm path"""
        args = arg.split()
        if len(args) != 1:
            print(self.do_rm.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        node = self._api.getNodeByPath(args[0], self.cwd)
        if node == None:
            print('Node not found: {}'.format(args[0]))
            return

        self._api.remove(node)


    def do_mv(self, arg):
        """Usage: mv srcpath dstpath"""
        args = arg.split()
        if len(args) != 2:
            print(self.do_mv.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        src_node = self._api.getNodeByPath(args[0], self.cwd)
        if src_node == None:
            print('{}: No such file or directory'.format(args[0]))
            return

        name = args[1]
        dst_node = self._api.getNodeByPath(name, self.cwd)
        if (dst_node != None) and (dst_node.getType() == MegaNode.TYPE_FILE):
            print('{}: Not a directory'.format(name))
            return

        if dst_node != None:
            self._api.moveNode(src_node, dst_node)
            return

        if '/' in name or '\\' in name:
            index1 = name.rfind('/')
            index2 = name.rfind('\\')
            index = None
            if index1 > index2:
                index = index1
            else:
                index = index2
            path = name[:index + 1]
            base = self._api.getNodeByPath(path, self.cwd)
            name = name[index + 1:]

            if base == None:
                print('{}: No such directory'.format(path))
                return

            if base.getType() == MegaNode.TYPE_FILE:
                print('{}: Not a directory'.format(path))
                return

            self._api.moveNode(src_node, base)
            if len(name) != 0:
                self._api.renameNode(src_node, name)
            return

        if dst_node == None:
            self._api.renameNode(src_node, name)
            return


    def do_pwd(self, arg):
        """Usage: pwd"""
        args = arg.split()
        if len(args) != 0:
            print(self.do_pwd.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        print('{} INFO: Current working directory: {}'
              .format(self.PROMPT, self._api.getNodePath(self.cwd)))


    def do_export(self, arg):
        """Usage: export path"""
        args = arg.split()
        if len(args) != 1:
            print(self.do_export.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        node = self._api.getNodeByPath(args[0], self.cwd)
        self._api.exportNode(node)


    def do_import(self, arg):
        """Usage: import exportedfilelink#key"""
        args = arg.split()
        if len(args) != 1:
            print(self.do_import.__doc__)
            return
        if self.cwd == None:
            print('INFO: Not logged in')
            return

        self._api.importFileLink(args[0], self.cwd)


    def do_whoami(self, arg):
        """Usage: whoami"""
        args = arg.split()
        if len(args) != 0:
            print(self.do_whoami.__doc__)
            return
        if not self._api.isLoggedIn():
            print('INFO: Not logged in')
            return
        print(self._api.getMyEmail())
        self._api.getAccountDetails()


    def do_passwd(self, arg):
        """Usage: passwd <old password> <new password> <new password>"""
        args = arg.split()
        if len(args) != 3:
            print(self.do_passwd.__doc__)
            return
        if not self._api.isLoggedIn():
            print('INFO: Not logged in')
            return

        if args[1] != args[2]:
            print('Mismatch, please try again')
            return

        self._api.changePassword(args[0], args[1])


    def do_quit(self, arg):
        """Usage: quit"""
        del self._api
        print('Bye!')
        return True


    def do_exit(self, arg):
        """Usage: exit"""
        del self._api
        print('Bye!')
        return True


if __name__ == '__main__':
    # Set up logging.
    logging.basicConfig(level=logging.INFO,
                        #filename='runner.log',
                        format='%(levelname)s\t%(asctime)s %(message)s')
    # Do the work.
    api = MegaApi('ox8xnQZL', None, None, 'Python megacli')
    shell = MegaShell(api)
    listener = AppListener(shell)
    api.addListener(listener)
    api = None
    shell.cmdloop()
    

File: examples/php/megacli.php
#!/usr/bin/env php

<?php

ini_set('display_errors', 'On');
error_reporting(E_ERROR);

set_time_limit(0);
set_include_path(get_include_path() . PATH_SEPARATOR . "../../bindings/php");

include 'megaapi.php';

if (file_exists('vendor/autoload.php'))
    require_once('vendor/autoload.php');
else
    require_once('Symfony/autoload.php');

use Symfony\Component\Console\Shell;
use Symfony\Component\Console\Application; 
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputOption;

$megaapi = NULL;
$cwd = NULL;

class AppListener implements MegaListenerInterface
{
	public function onRequestStart($megaApi, $request)
	{
	}
		
	public function onRequestFinish($megaApi, $request, $error)
	{
		global $cwd;
		
		if($error->getErrorCode() != MegaError::API_OK) 
		{	
                	print("INFO: Request finished with error ( " . $request . " )   Result: " .  $error . "\n");
			print("MEGA > ");
			return;
		}

		$requestType = $request->getType();
		if($requestType == MegaRequest::TYPE_LOGIN)
		{
			print("Fetchning nodes. Please wait...\n");
	                print("MEGA > ");
			$megaApi->fetchNodes();
		}
		else if($requestType == MegaRequest::TYPE_FETCH_NODES)
		{
                        print("Account correctly loaded\n");
                        print("MEGA > ");
			$cwd = $megaApi->getRootNode();
		}
		else if($requestType == MegaRequest::TYPE_EXPORT)
		{
			print("INFO: Exported link: " . $request->getLink() . "\n");
			print("MEGA > ");
		}
		else if($requestType == MegaRequest::TYPE_ACCOUNT_DETAILS)
		{
			$accountDetails = $request->getMegaAccountDetails();
			print("INFO: Account details received\n"); 
			print("Account e-mail: " . $megaApi->getMyEmail() . "\n");
			print("Storage: " . $accountDetails->getStorageUsed() . " of " . $accountDetails->getStorageMax() . 
				" (" . (100 * $accountDetails->getStorageUsed() / $accountDetails->getStorageMax()) . "%)\n");
			print("Pro level: " . $accountDetails->getProLevel() . "\n");
			print("MEGA > ");
		}
	}

	public function onRequestTemporaryError($megaApi, $request, $error)
	{	
		print("INFO: Request temporary error ( " . $request . " )   Error: " + $error . "\n");
	}
	
	public function onTransferStart($megaApi, $transfer)
	{
		print("INFO: Transfer start ( " . $transfer . " " . $transfer->getFileName() . " )\n");
	}
	
	public function onTransferFinish($megaApi, $transfer, $error)
	{		
		print("INFO: Transfer finished ( " . $transfer . " " . $transfer->getFileName() . " )   Result: " . $error . "\n");
		print("MEGA > ");
	}
	
	public function onTransferUpdate($megaApi, $transfer)
	{
		print("INFO: Transfer update ( " . $transfer . " " . $transfer->getFileName() . " )   Progress: " . $transfer->getTransferredBytes()/1024 ." KB of " . $transfer->getTotalBytes()/1024 . " KB, " . $transfer->getSpeed()/1024 . " KB/s\n");
	}
			
	public function onTransferTemporaryError($megaApi, $request, $error)
	{		
		print("INFO: Transfer temporary error ( " . $transfer . " " . $transfer->getFileName() . " )   Error: " . $error . "\n");
	}
	
	public function onUsersUpdate($megaApi, $users)
	{
	}

	public function onNodesUpdate($megaApi, $nodes)
	{
		global $cwd;

		if($nodes != NULL) 
		{
			print("INFO: Nodes updated ( " . count($nodes) . " )\n"); 
			print("MEGA > ");
		}
		else
			$cwd = $megaApi->getRootNode();
	}
	
	public function onReloadNeeded($megaApi)
	{
		
	}
}

class LoginCommand extends Command
{
    protected function configure()
    {
        $this->setName('login');
        $this->setDescription('Log in to a MEGA account');
        $this->addArgument('email', InputArgument::REQUIRED, 'Email of the account');
        $this->addArgument('password', InputArgument::REQUIRED, 'Password of the account');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		
        $email = $input->getArgument('email');
        $password = $input->getArgument('password');
		$megaapi->login($email, $password);
    }
}

class LogoutCommand extends Command
{
    protected function configure()
    {
        $this->setName('logout');
        $this->setDescription('Log out a MEGA account');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		$megaapi->logout();
		$cwd = null;
    }
}

class LsCommand extends Command
{
    protected function configure()
    {
        $this->setName('ls');
        $this->setDescription('List a MEGA folder');
        $this->addArgument('path', InputArgument::OPTIONAL, 'folder path');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$path = $input->getArgument('path');
        if ($path) 
        {
            $folder = $megaapi->getNodeByPath($path, $cwd);
        } 
        else 
        {
            $folder = $cwd;
        }


		$output->writeln("    .");
		if($megaapi->getParentNode($folder) != NULL)
		{
			$output->writeln("    ..");
		}

		$children = $megaapi->getChildren($folder);
		foreach($children as $node)
		{
			$output->write("    " . $node->getName());
			if($node->getType() == MegaNode::TYPE_FILE)
			{
				$output->writeln("   (" . $node->getSize() . " bytes)");
			}
			else
			{
				$output->writeln("   (folder)");
			}
		}		
    }
}

class MkdirCommand extends Command
{
    protected function configure()
    {
        $this->setName('mkdir');
        $this->setDescription('Create a folder');
        $this->addArgument('name', InputArgument::REQUIRED, 'folder name');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$name = $input->getArgument('name');
		$megaapi->createFolder($name, $cwd);
    }
}

class CdCommand extends Command
{
    protected function configure()
    {
        $this->setName('cd');
        $this->setDescription('Change the current directory');
        $this->addArgument('path', InputArgument::REQUIRED, 'new current directory');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$path = $input->getArgument('path');
		$new = $megaapi->getNodeByPath($path, $cwd);
		if($new == null)
		{
			$output->writeln("Invalid path");
			return;
		}
		
		$cwd = $new;
    }
}

class MvCommand extends Command
{
    protected function configure()
    {
        $this->setName('mv');
        $this->setDescription('Move a file/folder');
        $this->addArgument('source', InputArgument::REQUIRED, 'Source file/folder');
        $this->addArgument('destination', InputArgument::REQUIRED, 'Destination file/folder');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$source = $input->getArgument('source');
		$name = $input->getArgument('destination');

		$srcNode = $megaapi->getNodeByPath($source, $cwd);
		if($srcNode == null)
		{
			$output->writeln($source . ": No such file or directory");
			return;
		}
		
		$dstNode = $megaapi->getNodeByPath($name, $cwd);
		if(($dstNode != null) && $dstNode->isFile())
		{
			$output->writeln($name . ": Not a directory");
			return;
		}
		
		if($dstNode != null)
		{
			$megaapi->moveNode($srcNode, $dstNode);
			return;
		}
		
		if(strpos($name,"/") !== false || strpos($name,"\\") !== false)
		{		
			$str1 = strrchr($name, "/");
			$str2 = strrchr($name, "\\");
			$index = null;
			
			if($str2 == FALSE || strlen($str1) < strlen($str2))
			{	
				echo "A\n";
				$index = strlen($name) - strlen($str1);
			}
			else
			{
				echo "B\n";
				$index = strlen($name) - strlen($str2);
			}
			
			$path = substr($name, 0, $index);
			$base = $megaapi->getNodeByPath($path, $cwd);
			$name = substr($name, $index+1);	
			
			echo "INDEX: " . $index . "\n";
			echo "str1: " . $str1 . "\n";
			echo "str2: " . $str2 . "\n";
			echo "PATH: " . $path . "\n";
			echo "NAME: " . $name . "\n";
						
			if($base == null)
			{
				$output->writeln($path . ": Not such directory");
				return;
			}	
								
			if($base->isFile())
			{
				$output->writeln($path . ": Not a directory");
				return;
			}
			
			$megaapi->moveNode($srcNode, $base);
			if(strlen($name) != 0)
			{	
				$megaapi->renameNode($srcNode, $name);
			}
			return;
		}
		
		if($dstNode == null)
		{
			$megaapi->renameNode($srcNode, $name); 
		}
    }
}

class PwdCommand extends Command
{
    protected function configure()
    {
        $this->setName('pwd');
        $this->setDescription('Get the current working directory');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$output->writeln($megaapi->getNodePath($cwd));
    }
}

class RmCommand extends Command
{
    protected function configure()
    {
        $this->setName('rm');
        $this->setDescription('Remove a file/folder');
        $this->addArgument('path', InputArgument::REQUIRED, 'Path to file/folder to delete');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$path = $input->getArgument('path');
		$node = $megaapi->getNodeByPath($path, $cwd);
		if($node == null)
		{
			$output->writeln("Invalid path");
			return;
		}
			
		$megaapi->remove($node);
    }
}

class GetCommand extends Command
{
    protected function configure()
    {
        $this->setName('get');
        $this->setDescription('Download a file from MEGA');
        $this->addArgument('path', InputArgument::REQUIRED, 'Path to the file');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$path = $input->getArgument('path');
		$node = $megaapi->getNodeByPath($path, $cwd);
		if($node == null)
		{
			$output->writeln("Invalid path");
			return;
		}
		
		if(!$node->isFile())
		{
			$output->writeln("Not a file");
			return;
		}
			
		$megaapi->startDownload($node
		, "./"	 /*local path*/
		, null 	 /*custom name*/
		, null	 /*app data*/
		, false	 /*start first*/
		, null	 /*cancel token*/
		, 3      /*collision check COLLISION_CHECK_FINGERPRINT*/
		, 2      /*collision resolution COLLISION_RESOLUTION_NEW_WITH_N*/
		, null); /*listener*/
    }
}

class PutCommand extends Command
{
    protected function configure()
    {
        $this->setName('put');
        $this->setDescription('Upload a file to MEGA');
        $this->addArgument('path', InputArgument::REQUIRED, 'Path to the local file');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$path = $input->getArgument('path');
		$megaapi->startUpload($path
                ,$cwd           /*parent node*/
		, null  	/*filename*/
		, 0     	/*mtime*/
		, null  	/*appData*/
		, false 	/*isSourceTemporary*/
		, false 	/*startFirst*/
		, null		/*cancelToken*/
		, null);  	/*listener*/
    }
}

class ExportCommand extends Command
{
    protected function configure()
    {
        $this->setName('export');
        $this->setDescription('Generate a public link');
        $this->addArgument('path', InputArgument::REQUIRED, 'Path to the file/folder in MEGA');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$path = $input->getArgument('path');
		$node = $megaapi->getNodeByPath($path, $cwd);
		if($node == null)
		{
			$output->writeln("Invalid path");
			return;
		}
			
		$megaapi->exportNode($node);
    }
}

class ImportCommand extends Command
{
    protected function configure()
    {
        $this->setName('import');
        $this->setDescription('Import a MEGA public file link');
        $this->addArgument('link', InputArgument::REQUIRED, 'Public MEGA file link');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$link = $input->getArgument('link');
		$megaapi->importFileLink($link, $cwd);
    }
}

class PasswdCommand extends Command
{
    protected function configure()
    {
        $this->setName('passwd');
        $this->setDescription('Change the access password');
        $this->addArgument('current_password', InputArgument::REQUIRED, 'Current password');
        $this->addArgument('new_password', InputArgument::REQUIRED, 'New password');
        $this->addArgument('repeat_new_password', InputArgument::REQUIRED, 'New password');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$current_password = $input->getArgument('current_password');
		$new_password = $input->getArgument('new_password');
		$repeat_new_password = $input->getArgument('repeat_new_password');

		if($new_password != $repeat_new_password)
		{
			$output->writeln("Error: Password mismatch");
			return;
		}

		$megaapi->changePassword($current_password, $new_password);
    }
}

class WhoamiCommand extends Command
{
    protected function configure()
    {
        $this->setName('whoami');
        $this->setDescription('Show info about the current user');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$output->writeln($megaapi->getMyEmail());		
		$megaapi->getAccountDetails();
    }
}

class MountCommand extends Command
{
    protected function configure()
    {
        $this->setName('mount');
        $this->setDescription('Show incoming shares');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		global $megaapi;
		global $cwd;
		
		if($cwd == null)
		{
			$output->writeln("Not logged in");
			return;
		}
		
		$output->writeln("INFO: INSHARES:");
		$users = $megaapi->getContacts();
		foreach($users as $user)
		{
			$megaapi->getInShares();
			$inshares = $megaapi->getInShares($user);
			foreach($inshares as $share)
			{
				$output->writeln("INFO: INSHARE on " . $user->getEmail() . " " . $share->getName() . " Access level: " . $megaapi->getAccess($share));
			}
		}
    }
}

class ExitCommand extends Command
{
    protected function configure()
    {
        $this->setName('exit');
        $this->setDescription('Exit the app');
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
		exit(0);
	}
}

class QuitCommand extends ExitCommand
{
    protected function configure()
    {
        $this->setName('quit');
        $this->setDescription('Exit the app');
    }
}

MegaApi::setLogLevel(MegaApi::LOG_LEVEL_ERROR);
MegaApi::setLogToConsole(true);

$applistener = new AppListener();

$megaapi = new MegaApiPHP("API_KEY", "PHP megacli", ".");
$megaapi->addListener($applistener);

$application  = new Application('MEGA', 'PHP');
$application->add(new LoginCommand());
$application->add(new LogoutCommand());
$application->add(new LsCommand());
$application->add(new MkdirCommand());
$application->add(new CdCommand());
$application->add(new PwdCommand());
$application->add(new RmCommand());
$application->add(new GetCommand());
$application->add(new PutCommand());
$application->add(new ExitCommand());
$application->add(new QuitCommand());
$application->add(new ExportCommand());
$application->add(new ImportCommand());
$application->add(new WhoamiCommand());
$application->add(new PasswdCommand());
$application->add(new MountCommand());
$application->add(new MvCommand());

$shell = new Shell($application);
$shell->run();

?>


